internal/controller/plan/service_descriptor_utils.go

```diff
@@ -0,0 +1,141 @@
+/*
+Copyright (C) 2022-2023 ApeCloud Co., Ltd
+
+This file is part of KubeBlocks project
+
+This program is free software: you can redistribute it and/or modify
+it under the terms of the GNU Affero General Public License as published by
+the Free Software Foundation, either version 3 of the License, or
+(at your option) any later version.
+
+This program is distributed in the hope that it will be useful
+but WITHOUT ANY WARRANTY; without even the implied warranty of
+MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+GNU Affero General Public License for more details.
+
+You should have received a copy of the GNU Affero General Public License
+along with this program.  If not, see <http://www.gnu.org/licenses/>.
+*/
+
+package plan
+
+import (
+	"fmt"
+
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/types"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+
+	appsv1alpha1 "github.com/apecloud/kubeblocks/apis/apps/v1alpha1"
+	"github.com/apecloud/kubeblocks/internal/constant"
+	"github.com/apecloud/kubeblocks/internal/controller/builder"
+	"github.com/apecloud/kubeblocks/internal/controller/component"
+	intctrlutil "github.com/apecloud/kubeblocks/internal/controllerutil"
+)
+
+func GenServiceReferences(reqCtx intctrlutil.RequestCtx,
+	cli client.Client,
+	cluster *appsv1alpha1.Cluster,
+	clusterCompDef *appsv1alpha1.ClusterComponentDefinition,
+	clusterCompSpec *appsv1alpha1.ClusterComponentSpec,
+) (map[string]*appsv1alpha1.ServiceDescriptor, error) {
+	if cluster == nil || clusterCompDef == nil || clusterCompSpec == nil {
+		return nil, nil
+	}
+
+	if len(clusterCompDef.ServiceRefDeclarations) == 0 {
+		return nil, nil
+	}
+
+	serviceReferences := make(map[string]*appsv1alpha1.ServiceDescriptor, len(clusterCompDef.ServiceRefDeclarations))
+	for _, serviceRefDecl := range clusterCompDef.ServiceRefDeclarations {
+		for _, serviceRef := range clusterCompSpec.ServiceRefs {
```

Comment:
Handle the case that  the declaration is not defined

Reply:
> Handle the case that the declaration is not defined

I have marked this as TODO. The validation here should be processed based on whether it is actually referenced or not.

Reply:
Since the usage to referenced content may be implicit, KB cannot determine whether a reference is being used.

Reply:
Since the usage to referenced content may be implicit, KB cannot determine whether a reference is being used.